File: src/com/anathema_roguelike/main/utilities/AutoClassToInstanceMap.java
/*******************************************************************************
 * This file is part of AnathemaRL.
 *
 *     AnathemaRL is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     AnathemaRL is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with AnathemaRL.  If not, see <http://www.gnu.org/licenses/>.
 *******************************************************************************/
package com.anathema_roguelike.main.utilities;

import java.lang.reflect.InvocationTargetException;
import java.util.Collection;

import com.google.common.collect.ClassToInstanceMap;
import com.google.common.collect.MutableClassToInstanceMap;

public class AutoClassToInstanceMap<T> {
	
	private ClassToInstanceMap<T> classToInstanceMap = MutableClassToInstanceMap.create();
	
	public AutoClassToInstanceMap(Class<T> type, Class<?>[] parameterTypes, Object... args) {
		try {
			for(Class<? extends T> t : Utils.getSubclasses(type)) {
				classToInstanceMap.put(t, t.getConstructor(parameterTypes).newInstance(args));
			}
		} catch (InstantiationException | IllegalAccessException
				| IllegalArgumentException | InvocationTargetException
				| NoSuchMethodException | SecurityException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
			throw new RuntimeException();
		}
	}
	
	public AutoClassToInstanceMap(Class<T> type) {
		try {
			for(Class<? extends T> t : Utils.getSubclasses(type)) {
				classToInstanceMap.put(t, t.getConstructor().newInstance());
			}
		} catch (InstantiationException | IllegalAccessException
				| IllegalArgumentException | InvocationTargetException
				| NoSuchMethodException | SecurityException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
			throw new RuntimeException();
		}
	}
	
	public <I extends T> I get(Class<? extends I> key) {
		return classToInstanceMap.getInstance(key);
	}
	
	public Collection<T> getValues() {
		return classToInstanceMap.values();
	}
}

File: src/com/anathema_roguelike/main/utilities/AutoClassToInstanceMap.scala
/*******************************************************************************
 * Copyright (c) 2019. Carl Minden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package com.anathema_roguelike
package main.utilities

import com.anathema_roguelike.stats.characterstats.attributes.Agility

import scala.collection.mutable
import scala.reflect.ClassTag
import scala.reflect.runtime._
import scala.reflect.runtime.universe._
import com.anathema_roguelike.entities.characters.Character

class AutoClassToInstanceMap[T : TypeTag](parameterTypes: List[Class[_]], args: Any*) {

  private val classToInstanceMap = mutable.Map[Class[_ <: T], T]()

  Utils.getSubclasses[T]().foreach(t => {
    classToInstanceMap.put(t, t.getConstructor(parameterTypes.toArray:_*).newInstance(args.map(_.asInstanceOf[AnyRef]):_*))
  })

  def this() {
    this(List())
  }

  def get[I <: T : TypeTag]: I = classToInstanceMap(typeTagToClass[I]).asInstanceOf[I]

  def getValues: Iterable[T] = classToInstanceMap.values
}
